third_party/xls_go_math/fpexp_32.x

```diff
@@ -0,0 +1,244 @@
+//          if x >  7.09782712893383973096e+02 then exp(x) overflow
+//          if x < -7.45133219101941108420e+02 then exp(x) underflow
+//      [This is for the 64-bit implementation, whereas this dslx
+//       implementation is 32-bit]
+//
+// Constants:
+// The hexadecimal values are the intended ones for the following
+// constants. The decimal values may be used, provided that the
+// compiler will convert from decimal to binary accurately enough
+// to produce the hexadecimal values shown.
+// ====================================================
+
+// This file implements floating point exp,
+// which calcualtes e^x.
+// Note:
+//  - Input denormals are treated as/flushed to 0.
+//      (denormals-are-zero / DAZ).  Similarly,
+//      denormal results are flushed to 0.
+//  - No exception flags are raised/reported.
+//  - We emit a single, canonical representation for 
+//      NaN (qnan) but accept all NaN respresentations 
+//      as input
+
+import std
+import float32
+import third_party.xls_berkeley_softfloat.fpdiv_2x32
+import xls.modules.fpadd_2x32
+import xls.modules.fpldexp_32
+import xls.modules.fpmul_2x32
+import xls.modules.fpsub_2x32
+
+type F32 = float32::F32;
+
+
+const LN2HI = float32::unflatten(u32:0x3f317218);
+const LN2LO = float32::unflatten(u32:0x2f51cf7a);
+const LOG2E = float32::unflatten(u32:0x3fb8aa3b);
+
+const P1 = float32::unflatten(u32:0x3e2aaaab);
+const P2 = float32::unflatten(u32:0xbb360b61);
+const P3 = float32::unflatten(u32:0x388ab355);
+const P4 = float32::unflatten(u32:0xb5ddea0e);
+const P5 = float32::unflatten(u32:0x3331bb4c);
+
+// Returns e^(hi-lo) * 2^k, where |r| <= ln(2)/2
+fn expmulti(hi: F32, lo: F32, k: s32) -> F32 {
+  const ONE = float32::one(u1:0);
+  const TWO = F32{ sign: u1:0,
+                   bexp: float32::bias(u9:1),
+                   sfd: u23:0
+                  };
+
+  let range = fpsub_2x32::fpsub_2x32(hi, lo);
+  let range_sq = fpmul_2x32::fpmul_2x32(range, range);
+
+  // c := r - t*(P1+t*(P2+t*(P3+t*(P4+t*P5))))
+  let c = fpmul_2x32::fpmul_2x32(range_sq, P5);
+  let c = fpadd_2x32::fpadd_2x32(c, P4);
+  let c = fpmul_2x32::fpmul_2x32(range_sq, c);
+  let c = fpadd_2x32::fpadd_2x32(c, P3);
+  let c = fpmul_2x32::fpmul_2x32(range_sq, c);
+  let c = fpadd_2x32::fpadd_2x32(c, P2);
+  let c = fpmul_2x32::fpmul_2x32(range_sq, c);
+  let c = fpadd_2x32::fpadd_2x32(c, P1);
+  let c = fpmul_2x32::fpmul_2x32(range_sq, c);
+  let c = fpsub_2x32::fpsub_2x32(range, c);
+
+  // y := 1 - ((lo - (r*c)/(2-c)) - hi)
+  let numerator = fpmul_2x32::fpmul_2x32(range, c);
+  let divisor = fpsub_2x32::fpsub_2x32(TWO, c);
+  let div = fpdiv_2x32::fpdiv_2x32(numerator, divisor);
+  let y = fpsub_2x32::fpsub_2x32(lo, div);
+  let y = fpsub_2x32::fpsub_2x32(y, hi);
+  let y = fpsub_2x32::fpsub_2x32(ONE, y);
+
+  fpldexp_32::fpldexp_32(y, k)
+}
+
+// Returns e^x
+pub fn fpexp_32(x: F32) -> F32 {
+  const ZERO = float32::zero(u1:0);
+
+  // Flush subnormals.
+  let x = float32::zero(u1:0)
```

Comment:
we should probably have a flush_to_zero function somewhere. No need for you to do it though.

Reply:
Done (looks like it was already the float library).